traffic_ops/traffic_ops_golang/cdnfederation/cdnfederations.go

```diff
@@ -0,0 +1,389 @@
+
+		if id {
+			return nil, []error{tc.TenantDSUserNotAuthError}, tc.ForbiddenError
+		}
+
+		if yes, err := dbhelpers.CDNExists(parameters["name"], fed.ReqInfo.Tx); yes {
+			return federations, []error{}, tc.NoError
+		} else if err != nil { //internal server error
+			log.Errorf("verifying cdn exists: %v", err)
+			return nil, []error{tc.DBError}, tc.SystemError
+		} else { //the query ran as expected and the cdn does not exist
+			return nil, []error{errors.New("Resource not found.")}, tc.DataMissingError
+		}
+	}
+
+	return federations, []error{}, tc.NoError
+}
+
+func (fed *TOCDNFederation) Update() (error, tc.ApiErrorType) {
+
+	if ok, err := fed.isTenantAuthorized(); err != nil {
+		log.Errorf("checking tenacy: %v", err)
+		return tc.DBError, tc.SystemError
+	} else if !ok {
+		return tc.TenantUserNotAuthError, tc.ForbiddenError
+	}
+
+	//Deliveryservice IDs should not be included on update.
+	if fed.DeliveryServiceIDs != nil {
+		fed.DsId = nil
+		fed.XmlId = nil
+		fed.DeliveryServiceIDs = nil
+	}
+
+	result, err := fed.ReqInfo.Tx.NamedExec(updateQuery(), fed)
+	if err != nil {
+		return parseQueryError(err, "update")
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return tc.DBError, tc.SystemError
+	}
+
+	if rowsAffected != 1 {
+		if rowsAffected < 1 {
+			return errors.New("no federation found with this id"), tc.DataMissingError
+		}
+		return fmt.Errorf("this update affected too many rows: %d", rowsAffected), tc.SystemError
+	}
+
+	return nil, tc.NoError
+}
+
+//In the perl version, :name is ignored. It is not even verified whether or not
+//:name is a real cdn that exists. This mimicks the perl behavior.
+func (fed *TOCDNFederation) Delete() (error, tc.ApiErrorType) {
+
+	if ok, err := fed.isTenantAuthorized(); err != nil {
+		log.Errorf("checking tenacy: %v", err)
+		return tc.DBError, tc.SystemError
+	} else if !ok {
+		return tc.TenantUserNotAuthError, tc.ForbiddenError
+	}
+
+	log.Debugf("about to run exec query: %s with federation: %++v", deleteQuery(), fed)
+	result, err := fed.ReqInfo.Tx.NamedExec(deleteQuery(), fed)
+	if err != nil {
+		log.Errorf("received error: %++v from delete execution", err)
+		return tc.DBError, tc.SystemError
+	}
+
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		log.Errorf("getting rows affected: %v", err)
+		return tc.DBError, tc.SystemError
+	}
+
+	if rowsAffected != 1 {
+		if rowsAffected < 1 {
+			return errors.New("no federation with that id found"), tc.DataMissingError
+		} else {
+			return fmt.Errorf("this delete affected too many rows: %d", rowsAffected), tc.SystemError
+		}
+	}
+	return nil, tc.NoError
+}
+
+//Function not exported because although DELETE and UPDATE have normal tenacy check,
+//CREATE does not. No ds is associated on create. This isn't used for READ because
+//psql doesn't like nested queries.
+func (fed TOCDNFederation) isTenantAuthorized() (bool, error) {
+	if fed.ID == nil {
+		log.Errorf("unexpected nil id\n")
```

Comment:
It does return an error. The next line is `return nil, []error{tc.DBError}, tc.SystemError`. Maybe clarify?

Reply:
I don't see the line you're talking about. What I see is ```tenantID, err := getTenantIDFromFedID(*fed.ID, fed.ReqInfo.Tx.Tx)``` in which fed.ID could be nil, which would cause a nil pointer dereference panic

Reply:
I see it now.